packages/memory/src/index.ts

```diff
@@ -962,20 +947,156 @@ ${
   }
 
   /**
-   * Updates the metadata of a list of messages
+   * Updates content or other properties for a list of messages.
+   * Also updates embeddings if semantic recall is enabled.
    * @param messages - The list of messages to update
-   * @returns The list of updated messages
+   * @returns Promise that resolves to the list of updated messages
    */
   public async updateMessages({
     messages,
   }: {
-    messages: Partial<MastraMessageV2> & { id: string }[];
+    messages: Resolve<RequireOnly<MastraMessageV2, 'id'>>[];
   }): Promise<MastraMessageV2[]> {
     if (messages.length === 0) return [];
 
-    // TODO: Possibly handle updating the vector db here when a message is updated.
+    const config = this.getMergedThreadConfig();
+    try {
+      let embeddingTextsAndIds: ({
+        textForEmbedding: string;
+        vectorIds: string[];
+      } | null)[] = [];
+
+      let storedMessages: MastraMessageV2[] = [];
+      let storedMessagesById: Record<string, DeepPartial<MastraMessageV2>> = {};
+
+      if (config.semanticRecall) {
+        if (!this.vector) {
+          throw new Error(`Tried to update embeddings but this Memory instance doesn't have an attached vector db.`);
+        }
+        embeddingTextsAndIds = this.getEmbeddingTextAndIds(messages);
+        // fetch all passed messages to get stored vector chunk count
+        storedMessages = (
+          await this.storage.getMessages({
+            format: 'v2',
+            threadId: '',
+            selectBy: {
+              include: [
+                ...messages.map(msg => ({
+                  id: msg.id,
+                  threadId: msg.threadId,
+                })),
+              ],
+            },
+          })
+        ).flat(1);
+
+        storedMessagesById = storedMessages.reduce(
+          (acc: Record<string, (typeof messages)[number]>, msg) => Object.assign(acc, { [msg.id]: msg }),
+          {},
+        );
+      }
+
+      const updatedMessages = messages.map((message, i) => {
+        // remove createdAt so that storage.updateMessages doesn't invalidate stored dates
+        const { createdAt, ...rest } = message;
+        if (!message.content || !embeddingTextsAndIds[i]) return rest;
+        return deepMerge(rest, {
+          content: {
+            metadata: {
+              vectorIds: embeddingTextsAndIds[i].vectorIds,
+            },
+          },
+        });
+      });
+      const result = this.storage.updateMessages({ messages: updatedMessages });
+
+      if (embeddingTextsAndIds.length && this.vector) {
+        const vector = this.vector;
+        const promises: Promise<void | string[]>[] = [];
+        // update embeddings
+        embeddingTextsAndIds.forEach(async (el, i) => {
```

Comment:
I would rename `el`, not sure what that means, and use `embeddingIdx or something. Be more descriptive. Especially because you've defined `i` right above too.